pkg/antctl/raw/helper.go

```diff
@@ -59,40 +72,116 @@ func ResolveKubeconfig(cmd *cobra.Command) (*rest.Config, error) {
 	kubeconfig.Insecure = true
 	kubeconfig.CAFile = ""
 	kubeconfig.CAData = nil
-	if runtime.InPod {
-		if runtime.Mode == runtime.ModeAgent {
-			kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaAgentAPIPort))
-			kubeconfig.BearerTokenFile = agentapiserver.TokenPath
-		} else {
-			kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaControllerAPIPort))
-			kubeconfig.BearerTokenFile = controllerapiserver.TokenPath
-		}
+	if runtime.Mode == runtime.ModeAgent {
+		kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaAgentAPIPort))
+		kubeconfig.BearerTokenFile = agentapiserver.TokenPath
+	} else {
+		kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaControllerAPIPort))
+		kubeconfig.BearerTokenFile = controllerapiserver.TokenPath
 	}
 }
 
-func CreateAgentClientCfg(k8sClientset kubernetes.Interface, antreaClientset antrea.Interface, cfgTmpl *rest.Config, nodeName string) (*rest.Config, error) {
-	node, err := k8sClientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+// this is useful for testing
+var runCommandFromPodWrapper = runCommandFromPod
+
+func runCommandFromPod(ctx context.Context, client kubernetes.Interface, kubeconfig *rest.Config, podNamespace string, podName string, containerName string, cmd []string) (stdout []byte, stderr []byte, err error) {
+	request := client.CoreV1().RESTClient().Post().
+		Namespace(podNamespace).
+		Resource("pods").
+		Name(podName).
+		SubResource("exec").
+		Param("container", containerName).
+		VersionedParams(&corev1.PodExecOptions{
+			Command: cmd,
+			Stdin:   false,
+			Stdout:  true,
+			Stderr:  true,
+			TTY:     false,
+		}, scheme.ParameterCodec)
+	exec, err := remotecommand.NewSPDYExecutor(kubeconfig, "POST", request.URL())
 	if err != nil {
-		return nil, fmt.Errorf("error when looking up Node %s: %w", nodeName, err)
+		return nil, nil, err
+	}
+	var stdoutB, stderrB bytes.Buffer
+	if err := exec.StreamWithContext(ctx, remotecommand.StreamOptions{
+		Stdout: &stdoutB,
+		Stderr: &stderrB,
+	}); err != nil {
+		return stdoutB.Bytes(), stderrB.Bytes(), err
+	}
+	return stdoutB.Bytes(), stderrB.Bytes(), nil
+}
+
+// GetAgentCert retrieves the self-signed certificate generated by the Agent on startup and written
+// to its local filesystem. Retrieving this certificate by exec'ing into the Pod is the only way to
+// establish a direct secure connection to the Agent.
+func GetAgentCert(ctx context.Context, client kubernetes.Interface, kubeconfig *rest.Config, podNamespace string, podName string) ([]byte, error) {
+	// use path instead of path/filepath as the target system is different from the currrent one.
+	certPath := path.Join(agentapiserver.CertDirectory, agentapiserver.CertPairName+".crt")
+	cmd := []string{"cat", certPath}
+	stdout, _, err := runCommandFromPodWrapper(ctx, client, kubeconfig, podNamespace, podName, "antrea-agent", cmd)
```

Comment:
Do you think it makes sense to publish the CA certificate to a field of AntreaAgentInfo? Then antctl could get APIPort and CACertificate from the same place, and may be helpful when other clients want to connect agent API securely without requiring the Pod exec permission.

Reply:
I think it's not a bad idea. Let me think about it a bit more, and I'll make the change tomorrow.

Reply:
I went ahead with that change, thanks for the suggestion.
It is better as it enables support for non-Pod use cases (Windows, Baremetal).
My one concern is that it increases the size of the CRs. I am not sure if it can eventually become an issue for large clusters with many Nodes. As a reminder, we call `Update` every minute to update the CR. But there is a solution we could consider: using `Patch` instead of `Update` so that data that is static / has not changed does not need to be sent to the APIServer. I am not sure whether this is something that has been discussed before?

Reply:
I think it should be fine. A CA bundle's size is about 1,460 bytes. Even in a cluster of 1000 Nodes, it just increases the storage size 1,460 KB * N (a couple of history versions).
Using Patch to update AntreaAgentInfo makes sense to me. I don't remember if we discussed it before, but it was not so important previously when it just contained a few small fields.

Reply:
It's twice that (about 3KB) with the current bundle (cert + CA) but your observation still holds.